File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/deserializers/AdditivesWrapperDeserializer.java
package openfoodfacts.github.scrachx.openfood.network.deserializers;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import openfoodfacts.github.scrachx.openfood.models.AdditiveResponse;
import openfoodfacts.github.scrachx.openfood.models.AdditivesWrapper;

/**
 * Created by Lobster on 03.03.18.
 */

public class AdditivesWrapperDeserializer extends StdDeserializer<AdditivesWrapper> {


    private static final String EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY = "efsa_evaluation_overexposure_risk";
    private static final String EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI = "efsa_evaluation_exposure_95th_greater_than_adi";
    private static final String EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL = "efsa_evaluation_exposure_95th_greater_than_noael";
    private static final String EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI = "efsa_evaluation_exposure_mean_greater_than_adi";
    private static final String EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL = "efsa_evaluation_exposure_mean_greater_than_noael";

    public AdditivesWrapperDeserializer() {
        super(AdditivesWrapper.class);
    }

    @Override
    public AdditivesWrapper deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
        List<AdditiveResponse> additives = new ArrayList<>();

        JsonNode mainNode = jp.getCodec().readTree(jp);
        Iterator<Map.Entry<String, JsonNode>> mainNodeIterator = mainNode.fields();

        while (mainNodeIterator.hasNext()) {
            final Map.Entry<String, JsonNode> subNode = mainNodeIterator.next();
            JsonNode namesNode = subNode.getValue().get(DeserializerHelper.NAMES_KEY);

            if (namesNode != null) {
                Map<String, String> names = DeserializerHelper.extractNames(namesNode);

                String overexposureRisk = null;
                String exposureMeanGreaterThanAdi = null;
                String exposureMeanGreaterThanNoael = null;
                String exposure95ThGreaterThanAdi = null;
                String exposure95ThGreaterThanNoael = null;
                if( subNode.getValue().has( EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY ) )
                {
                    // parse the overexposure risk the default value is "no"
                    overexposureRisk = subNode.getValue().get( EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY ).
                            get( DeserializerHelper.EN_KEY ).textValue().replaceFirst("^en:", "");

                    // update exposure evaluation map
                    if( subNode.getValue().has( EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI ) )
                    {
                        exposureMeanGreaterThanAdi = subNode.getValue().get( EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI ).get( DeserializerHelper.EN_KEY ).textValue();
                    }

                    if( subNode.getValue().has( EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL ) )
                    {
                        exposureMeanGreaterThanNoael = subNode.getValue().get( EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL ).get( DeserializerHelper.EN_KEY ).textValue();
                    }

                    if( subNode.getValue().has( EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI ) )
                    {
                        exposure95ThGreaterThanAdi = subNode.getValue().get( EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI ).get( DeserializerHelper.EN_KEY ).textValue();
                    }

                    if( subNode.getValue().has( EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL ) )
                    {
                        exposure95ThGreaterThanNoael = subNode.getValue().get( EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL ).get( DeserializerHelper.EN_KEY ).textValue();
                    }
                }

                AdditiveResponse additiveResponse;
                if (subNode.getValue().has(DeserializerHelper.WIKIDATA_KEY)) {
                    additiveResponse = new AdditiveResponse(subNode.getKey(), names, overexposureRisk, subNode.getValue().get(DeserializerHelper.WIKIDATA_KEY).toString());
                } else {
                    additiveResponse = new AdditiveResponse(subNode.getKey(), names, overexposureRisk);
                }
                additiveResponse.setExposureEvalMap( exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael );
                additives.add(additiveResponse);
            }
        }


        AdditivesWrapper wrapper = new AdditivesWrapper();
        wrapper.setAdditives(additives);

        return wrapper;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/additive/AdditivesWrapperDeserializer.kt
package openfoodfacts.github.scrachx.openfood.models.entities.additive

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import openfoodfacts.github.scrachx.openfood.utils.DeserializerHelper
import openfoodfacts.github.scrachx.openfood.utils.DeserializerHelper.extractMapFromJsonNode
import java.io.IOException

/**
 * Created by Lobster on 03.03.18.
 */
class AdditivesWrapperDeserializer : StdDeserializer<AdditivesWrapper>(AdditivesWrapper::class.java) {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, ctxt: DeserializationContext): AdditivesWrapper {
        val additives = mutableListOf<AdditiveResponse>()
        val mainNode = jp.codec.readTree<JsonNode>(jp)
        val mainNodeIterator = mainNode.fields()
        while (mainNodeIterator.hasNext()) {
            val subNode = mainNodeIterator.next()
            val namesNode = subNode.value[DeserializerHelper.NAMES_KEY]
            if (namesNode != null) {
                val names = extractMapFromJsonNode(namesNode)
                var overexposureRisk: String? = null
                var exposureMeanGreaterThanAdi: String? = null
                var exposureMeanGreaterThanNoael: String? = null
                var exposure95ThGreaterThanAdi: String? = null
                var exposure95ThGreaterThanNoael: String? = null
                if (subNode.value.has(EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY)) {
                    // parse the overexposure risk the default value is "no"
                    overexposureRisk = subNode.value[EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY][DeserializerHelper.EN_KEY].textValue().replaceFirst("^en:".toRegex(), "")

                    // update exposure evaluation map
                    if (subNode.value.has(EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI)) {
                        exposureMeanGreaterThanAdi = subNode.value[EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI][DeserializerHelper.EN_KEY].textValue()
                    }
                    if (subNode.value.has(EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL)) {
                        exposureMeanGreaterThanNoael = subNode.value[EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL][DeserializerHelper.EN_KEY].textValue()
                    }
                    if (subNode.value.has(EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI)) {
                        exposure95ThGreaterThanAdi = subNode.value[EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI][DeserializerHelper.EN_KEY].textValue()
                    }
                    if (subNode.value.has(EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL)) {
                        exposure95ThGreaterThanNoael = subNode.value[EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL][DeserializerHelper.EN_KEY].textValue()
                    }
                }
                val additiveResponse = if (subNode.value.has(DeserializerHelper.WIKIDATA_KEY)) {
                    AdditiveResponse(subNode.key, names, overexposureRisk, subNode.value[DeserializerHelper.WIKIDATA_KEY].toString())
                } else {
                    AdditiveResponse(subNode.key, names, overexposureRisk)
                }
                additiveResponse.setExposureEvalMap(exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael)
                additives.add(additiveResponse)
            }
        }
        return AdditivesWrapper(additives.toList())
    }

    companion object {
        private const val EFSA_EVALUATION_OVEREXPOSURE_RISK_KEY = "efsa_evaluation_overexposure_risk"
        private const val EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_ADI = "efsa_evaluation_exposure_95th_greater_than_adi"
        private const val EFSA_EVALUATION_EXPOSURE_95TH_GREATER_THAN_NOAEL = "efsa_evaluation_exposure_95th_greater_than_noael"
        private const val EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_ADI = "efsa_evaluation_exposure_mean_greater_than_adi"
        private const val EFSA_EVALUATION_EXPOSURE_MEAN_GREATER_THAN_NOAEL = "efsa_evaluation_exposure_mean_greater_than_noael"
    }
}